### DISCUSSION TITLE
Globs as Sources in Extension Modules

### DISCUSSION THREAD

**Discussion by codeandfire** [pypa/setuptools#3384] [Q&A] [answered]
Hi,

I just noticed that while configuring extension modules with Setuptools, we cannot use glob patterns in the list of `sources`. Is this intentional?

For example, if we have a package like this
```
mypkg/
    foo.c
    bar.c
    setup.py
```
and then in `setup.py`:
```python
from setuptools import setup, Extension

foobar_ext = Extension(
    name='foobar',
    sources=['*.c'])

setup(ext_modules=[foobar_ext])
```
Now if we run `pip install .` inside `mypkg`, this gives the following error:
```
running build_ext
building 'foobar' extension
gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -flto=auto -ffat-lto-objects -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -flto=auto -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -flto=auto -fPIC -I/home/codeandfire/tmp/.venv/include -I/usr/include/python3.10 -c *.c -o build/temp.linux-x86_64-cpython-310/*.o
cc1: fatal error: *.c: No such file or directory
compilation terminated.
error: command '/usr/bin/gcc' failed with exit code 1
```
and it seems this goes only when we explicitly list out the C source code files:
```python
foobar_ext = Extension(
    name='foobar',
    sources=['foo.c', 'bar.c'])
```

Would like to understand this better. Thanks!

**Comment by abravalheri** [comment, marked as answer]
Hi @codeandfire, I am not familiar with this part of the code yet, but I think this is originally intentional.

Since right now to add extensions you need to use `setup.py`, the assumption might have been that if the user need to expand globs, they can do it themselves using the standard library.

In the future we might want to add extensions support to `pyproject.toml`, and then have globs expanded in the case of config via TOML, but I haven't arrived there yet...

**Reply by codeandfire** [reply]
Ah I see, thank you very much @abravalheri!
